cproxy.1
.TH cproxy 1 "1.0" "Internet Connection, Inc." "Ad-Hoc Clusters"
.SH NAME
cproxy \- logically join two clusters
.SH SYNOPSIS
.B cproxy self target

Relays announcements between
.B $GROUP
and
.B target
as well as rewrites source/destination addresses to pass through cproxy.
.SH DESCRIPTION
.B cproxy
is useful for launching man-in-the-middle attacks against
.I cexec(1)
but it's also useful for proxying jobs over a wide-area network.

.B cproxy
attempts to detect loops within itself, but running two copies of
.B cproxy
on the same network in opposite directions can easily eat a large
chunk of bandwidth. Don't do it. Use DMZ or mrouted if possible.
.SH ENVIRONMENT
.B $GROUP
\- contains the "group name" - an IP address (possibly multicast) representing
the address to be used for receiving job requests.
.I Default: 255.255.255.255
.SH EXIT CODES
.B cproxy
normally doesn't exit. All exit codes are errors.
.SH SEE ALSO
ckeygen(1), cexec(1), cservice(1), crat(1)


CHANGELOG
2008-08-07 18:02  geocar

	* Makefile, bio.c, build-cross.sh, cexec.c, ckeygen.c, cproxy.c,
	  crat.c, cservice.c, deploy.sh, hash.c, keyfile.c, load.c,
	  multipipe.c, netnum.c, netnum.h, parity.c, sha1.c, sha1.h,
	  strline.c, strline.h, types.h: port to 64bit system

2007-03-02 22:05  geocar

	* deploy.sh: amd64

2007-01-30 21:27  geocar

	* cexec.c: wee

2006-12-08 22:48  geocar

	* bio.c, cexec.c, cservice.c, deploy.sh: moving it around...

2006-12-08 17:48  geocar

	* bio.c, build-cross.sh, cexec.c, cio.c, cservice.c, deploy.sh: fix
	  base when doing copyout; dont close on write (sigpipe helps
	  noone) and sparc is dead. oh, and nul-group is all-group

2006-07-31 06:46  geocar

	* cservice.c: support $BOOST mode

2006-07-31 06:46  geocar

	* load.c: use gnu load

2006-06-23 06:59  geocar

	* cexec.c: dont regenerate parity (we cannot remember long enough)

2006-06-23 06:59  geocar

	* cservice.c: fix delays, and make pubkey toplevel so we can test

2006-06-23 05:27  geocar

	* Makefile: uses sockop

2006-06-23 05:27  geocar

	* cexec.c: add testing feature, and skipto to try next-best host

2006-06-23 05:26  geocar

	* cservice.c: check signature for safety and use sockop_nonblock

2006-06-23 05:26  geocar

	* hash.c: zero s1/s2 barrels

2006-06-23 05:26  geocar

	* keyfile.c: check zero length vlpoint

2006-06-23 05:26  geocar

	* cio.c, sockop.c: bleh, convenience i suppose

2005-11-08 07:51  geocar

	* cexec.c: remove did_io optimization; it's complicated

2005-03-17 10:07  geocar

	* ckeygen.1: adding documentation about -stdin flag

2005-03-17 10:06  geocar

	* ckeygen.c: generate keys based on stdin with -stdin option now...

2005-03-17 10:06  geocar

	* cexec.c: did_io when writing too...

2005-03-01 12:58  geocar

	* cservice.c: memalloc isn't fatal PEER_vars, don't attempt to dup2
	  a -1 fd (bitmap)

2005-03-01 11:31  geocar

	* cproxy.c: don't waste file descriptors all wily-nilly, also
	  multihomed SO_BINDTODEVICE capable boxes are now assisted...

2005-03-01 11:31  geocar

	* cexec.c: put lseek hack back in

2005-02-17 08:11  geocar

	* parity.c: use urandom instead of random; the lifetime of the
	  check-bits are so short that other attacks are more likely to be
	  useful

2005-02-17 07:59  geocar

	* cservice.c: do bit flipping if near or far is running on a really
	  strange platform

2005-02-17 07:58  geocar

	* cexec.c: return didio hack

2005-02-17 07:58  geocar

	* cexec.c, cluster.h: move bytesex constant out of cexec

2005-02-17 07:57  geocar

	* bio.c: sink output in blast and copy

2005-02-17 06:27  geocar

	* LICENSE, Makefile, cexec.c, cio.c, cio.h, cservice.c, keyfile.c:
	  moved cread/cwrite out of keyfile and changed write-hack to use
	  4bytes instead of a vlpoint. the write-hack might prove useful
	  someday...

2005-02-16 14:40  geocar

	* cexec.c, cservice.c: turn keepalive back on

2005-02-16 13:02  geocar

	* cexec.c, cservice.c: don't use keepalive anymore, protect
	  main-channel for sanity, and add an extra vlpoint-exchange just
	  to force data both ways on the socket before working/forking

2005-02-16 13:00  geocar

	* bio.c: protect fcntl changes

2005-02-16 12:50  geocar

	* bio.c: turn off nonblock when blasting, and honor EAGAIN on
	  bio_more

2005-02-16 12:47  geocar

	* keyfile.c: turn off nonblock when in cread/cwrite

2005-02-16 09:54  geocar

	* cservice.1, cservice.c: set $REMOTE_ADDR, $REMOTE_PORT,
	  $SERVER_ADDR, $SERVER_PORT so cservice child can do things like
	  ident-check the cexec peer if it likes

2005-02-16 09:53  geocar

	* crat.c: unknown keys are listed in hex for brevity

2005-02-16 09:53  geocar

	* strline.c, strline.h: support for hex numbers too

2005-02-16 08:29  geocar

	* LICENSE, Makefile, cexec.c, crat.c, cservice.c, fdm.c, pair.c,
	  sockop.c, sockop.h: move sockops into a separate file

2005-02-16 08:28  geocar

	* cproxy.c: bugfix; detect hangup on write; move expires operation
	  before gathering fds so cproxy doesn't bomb

2005-02-15 10:44  geocar

	* cexec.c: restart processing if we haven't done any I/O but still
	  got an error,

2005-02-15 10:44  geocar

	* bio.c: treat EINTR and EAGAIN the same

2005-02-15 10:43  geocar

	* cservice.c: map fd masks correctly; increased timeouts, don't try
	  and read from wronly fds, use PIPE_BUF for buffers, and send
	  close alerts for write-errors

2005-02-15 10:42  geocar

	* fdm.c: map rdonly file descriptors correctly now; treat ttys
	  specially

2005-02-15 10:41  geocar

	* keyfile.c: treat EINTR and EAGAIN the same; also be pendantic
	  about short != 16 bits

2005-02-14 17:09  geocar

	* cexec.c: use a different (better?) announcement schedule

2005-02-14 17:09  geocar

	* cproxy.c: un-nonblock fds, turn on keepalive/ndelay on both
	  sides, honor EAGAIN

2005-02-14 17:09  geocar

	* cservice.1, cservice.c: lower default concurrency, drop
	  unnecessary checks, and recover from autorat failures faster

2005-02-14 16:35  geocar

	* load.c: system_load() isn't required to actually function

2005-02-14 11:01  geocar

	* cexec.1, ckeygen.1, cproxy.1, crat.1, cservice.1: adding
	  documentation

2005-02-14 08:57  geocar

	* cexec.c: don't waste an fd

2005-02-14 08:47  geocar

	* cexec.c, cservice.c: added pingpong detection, configurable
	  timeouts, configurable port ranges, using simplexes instead of
	  duplexes where possible, sending fd usage map, and debugging
	  nonsense for cexec

2005-02-14 08:43  geocar

	* crat.c, strline.c, strline.h: move string opers stuff

2005-02-14 08:42  geocar

	* fdm.c: fixes in detecting file descriptors on Linux

2005-02-14 08:32  geocar

	* pair.c: turn in so_linger on our pair links.

2005-02-14 08:30  geocar

	* parity.c: use /dev/random if available

2004-12-16 07:21  geocar

	* cservice.c: moving waitpid chunk earlier,change autorat timeouts,
	  and disable need_delay after delay (procrastinating autorats)

2004-12-16 06:37  geocar

	* cservice.c: default maxchildren now 100

2004-12-16 06:31  geocar


cproxy.c
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <fcntl.h>

#include "cluster.h"

#define CHATTER_TIMEOUT	120
#define BRINGUP_TIMEOUT	30

unsigned char buffer[32768];
unsigned int bufferx = 0;

static int bufferin(int fd)
{
	int r;

	if (fd == -1) return 0;
	do {
		r = read(fd, buffer, sizeof(buffer));
	} while (r == -1 && errno == EINTR);
	if (r == -1 && errno == EAGAIN) {
		bufferx = 0;
		return 1;
	}
	if (r < 1) return 0;
	bufferx = r;
	return 1;
}
static int bufferout(int fd)
{
	int r, x;

	if (fd == -1) return 0;
	for (x = 0; x < bufferx;) {
		do {
			r = write(fd, buffer+x, bufferx-x);
		} while (r == -1 && (errno == EINTR || errno == EAGAIN));
		if (r < 1) return 0; /* blech */
		x += r;
	}
	return 1;
}


struct cpipe {
	int left, right;
	struct sockaddr_in todo;
	time_t expires;

	struct cpipe *next;
	int state;
};

static struct cpipe *top = 0;

static struct cpipe *add_cpipe(int fd, struct sockaddr_in *ct)
{
	struct cpipe *x;
	int j;

	/* check cluster list and don't create a new cpipe; instead return
	 * an existing one...
	 */
	for (x = top; x; x = x->next) {
		if (x->left == -1) continue;
		if (x->todo.sin_family != ct->sin_family) continue;
		if (x->todo.sin_port != ct->sin_port) continue;
		if (memcmp(&x->todo.sin_addr, &ct->sin_addr,
					sizeof(ct->sin_addr)) != 0) continue;
		/* found a dup */

		/* already accepted/connected */
		if (x->state != 0) return 0;

		/* increase timeout */
		x->expires = time(0) + BRINGUP_TIMEOUT;
		return x;
	}


	x = (struct cpipe *)malloc(sizeof(struct cpipe));
	if (!x) return 0;

	x->left = fd;
	x->right = -1;
	memcpy(&x->todo, ct, sizeof(struct sockaddr_in));
	x->expires = time(0) + BRINGUP_TIMEOUT;

	x->state = 0;
	x->next = top;
	top = x;

	return x;
}
static int select_cpipes(int maxfd, fd_set *r, fd_set *w)
{
	register struct cpipe *c;

	for (c = top; c; c = c->next) {
		/* busy state */
		if (c->left > -1 && c->state != 2) {
			FD_SET(c->left, r);
			if (c->left > maxfd) maxfd = c->left;
		}
		if (c->right > -1 && c->state != 3) {
			if (c->state == 1) {
				FD_SET(c->right, r);
			} else {
				FD_SET(c->right, w);
			}
			if (c->right > maxfd) maxfd = c->right;
		}
	}
	return maxfd;
}
static void accept_cpipe(struct cpipe *c)
{
	struct sockaddr_in sin;
	int fd, r, j;

	j = sizeof(sin);
	r = accept(c->left, (struct sockaddr*)&sin, &j);
	if (r == -1) return;
	fd = socket(AF_INET, SOCK_STREAM, 0);
	if (fd == -1) {
		(void)close(r);
		return;
	}
	sockop_nonblock(fd, 1);
	sockop_keepalive(fd, 1);
	sockop_ndelay(fd, 1);

	sockop_nonblock(r, 1);
	sockop_keepalive(r, 1);
	sockop_ndelay(r, 1);
	if (connect(fd, (struct sockaddr *)&c->todo, sizeof(c->todo)) == 0) {
		/* immediately successful! */
		(void)close(c->left);
		c->left = r;
		c->right = fd;
		c->state = 1;
		sockop_nonblock(fd, 0);
		sockop_nonblock(r, 0);
		return;
	}

#ifdef EINPROGRESS
	if (errno != EINPROGRESS)
#endif
#ifdef EWOULDBLOCK
		if (errno != EWOULDBLOCK)
#endif
		{
			(void)close(fd);
			return;
		}

	(void)close(c->left);
	c->left = r;
	c->right = fd;
	c->expires = time(0) + BRINGUP_TIMEOUT;
	return;
}

static void update_cpipes(fd_set *r, fd_set *w)
{
	register struct cpipe *c;
	struct sockaddr_in sin;
	int dummy;

	for (c = top; c; c = c->next) {
		if (c->left == -1) continue;
		if (FD_ISSET(c->left, r)) {
			if (c->state == 0) {
				accept_cpipe(c);
			} else if (c->state == 1 || c->state == 3) {
				if (bufferin(c->left)) {
					c->expires = time(0) + CHATTER_TIMEOUT;
					if (!bufferout(c->right)) {
						c->expires = time(0) + CHATTER_TIMEOUT;
						c->state = 2;
						/* hangup */
						(void)shutdown(c->right, SHUT_WR);
						(void)shutdown(c->left, SHUT_RD);
					}
				} else if (c->state == 3) {
					(void)close(c->right);
					(void)close(c->left);
					c->left = c->right = -1;
					c->state = 0;
				} else {
					c->expires = time(0) + CHATTER_TIMEOUT;
					c->state = 2;
					/* hangup */
					(void)shutdown(c->right, SHUT_WR);
					(void)shutdown(c->left, SHUT_RD);
				}
			}
		}
		if (c->right == -1) continue;
		if (c->state == 0) {
			if (FD_ISSET(c->right, w)) {
				dummy = sizeof(sin);
				if (getpeername(c->right,
						(struct sockaddr *)&sin,
						&dummy) == -1) {
					/* failure */
					(void)close(c->right);
					(void)close(c->left);
					c->left = c->right = -1;
				} else {
					/* actually connected! hurrah! */
					c->state = 1;
					sockop_nonblock(c->left, 0);
					sockop_nonblock(c->right, 0);
					c->expires = time(0) + CHATTER_TIMEOUT;
				}
			}

		} else if (FD_ISSET(c->right, r)) {
			if (c->state == 1 || c->state == 2) {
				if (bufferin(c->right)) {
					c->expires = time(0) + CHATTER_TIMEOUT;
					if (!bufferout(c->left)) {
						c->expires = time(0) + CHATTER_TIMEOUT;
						c->state = 3;
						/* hangup */
						(void)shutdown(c->left, SHUT_WR);
						(void)shutdown(c->right, SHUT_RD);
					}
				} else if (c->state == 2) {
					(void)close(c->right);
					(void)close(c->left);
					c->left = c->right = -1;
					c->state = 0;
				} else {
					c->expires = time(0) + CHATTER_TIMEOUT;
					c->state = 3;
					/* hangup */
					(void)shutdown(c->left, SHUT_WR);
					(void)shutdown(c->right, SHUT_RD);
				}
			}
		}
	}
}
static int expire_cpipes(void)
{
	register int min = -1;
	register struct cpipe *c, *cl;
	register int df;
	time_t now;

	time(&now);
	for (c = top, cl = 0; c; c = c->next) {
		df = c->expires - now;
		if (df > 0 && (c->left > -1 || c->right > -1)) {
			if (min == -1 || min > df) min = df;
			cl = c;
		} else {
			/* reap */
			if (cl) {
				cl->next = c->next;
			} else {
				top = c->next;
			}
			if (c->left > -1) (void)close(c->left);
			if (c->right > -1) (void)close(c->right);
			(void)free(c);
		}
	}

	return min;
}
static int lookup_peer(int fd)
{
	register struct cpipe *c;
	for (c = top; c; c = c->next) {
		if (c->left == fd) {
			if (c->state == 0) return -1;
			return c->right;
		}
		if (c->right == fd) {
			if (c->state == 0) return -1;
			return c->left;
		}
	}
	return -1;
}

static char *hasdev(char **xret)
{
	char *x, *y;

	x = *xret;
	y = (char *)strchr(x, ':');
	if (!y) return 0;
	(*xret) = (y+1);
	return x;
}

int main(int argc, char *argv[])
{
	struct cpipe *c;
	unsigned char packet[512];
	struct timeval tv, *ptv;
	struct sockaddr_in sin, sendt, binda;
	struct ip_mreq gamr;
	in_addr_t ina, inb;
	unsigned int bf;
	fd_set rfds, wfds;
	char *q, *p;
	int bfd, sfd, maxfd, fd;
	int i, r;

	if (argc != 3 && argc != 2) {
#define M "Usage: cproxy bindout [sendto]\n"
		for (i = 0; i < sizeof(M)-1;) {
			do {
				r = write(2, M+i,(sizeof(M)-1)-i);
			} while (r == -1 && errno == EINTR);
			if (r < 1) exit(255);
			i += r;
		}
		exit(0);
#undef M
	}

	bfd = socket(AF_INET, SOCK_DGRAM, 0);
	sfd = socket(AF_INET, SOCK_DGRAM, 0);
	if (bfd == -1 || sfd == -1) exit(111);

	memset(&sin, 0, sizeof(sin));
	sin.sin_family = AF_INET;

	q = (char *)getenv("GROUP");
	if (!q) q = "255.255.255.255";
	if ((p = hasdev(&q))) {
#ifdef SO_BINDTODEVICE
		(void)setsockopt(bfd, SOL_SOCKET, SO_BINDTODEVICE,
				 p, strlen(p));
#endif
	}
	
	ina = inet_addr(q);
	memcpy(&sin.sin_addr, &ina, sizeof(in_addr_t));
	i = *((unsigned char *)&ina);
	if (i >= 224 && i < 255) {
		/* multicast */
		i = 1; (void)setsockopt(bfd, IPPROTO_IP, IP_MULTICAST_LOOP, &i, sizeof(i));
		i = 31; (void)setsockopt(bfd, IPPROTO_IP, IP_MULTICAST_TTL, &i, sizeof(i));
	}
	sockop_broadcast(bfd, 1);
	sockop_reuseaddr(bfd, 1);
	sin.sin_port = htons(CLUSTER_PORT);
	if (bind(bfd, (struct sockaddr *)&sin, sizeof(sin)) == -1) exit(111);

	/* sending for broadcasts... */
	memset(&sin, 0, sizeof(sin));
	sin.sin_family = AF_INET;

	inb = inet_addr(argv[1]);
#ifndef INADDR_NONE
#define INADDR_NONE	((in_addr_t)-1)
#endif
	if (inb == INADDR_NONE) exit(111);
	memcpy(&sin.sin_addr, &inb, sizeof(in_addr_t));
	r = *((unsigned char *)&inb);
	if (r >= 224 && r < 255) exit(255); /* can't fucking bind multicast */

	q = (argc == 2 ? "255.255.255.255" : argv[2]);
	if ((p = hasdev(&q))) {
#ifdef SO_BINDTODEVICE
		(void)setsockopt(sfd, SOL_SOCKET, SO_BINDTODEVICE,
				 p, strlen(p));
#endif
	}

	ina = inet_addr(q);
	i = *((unsigned char *)&ina);
	if (i >= 224 && i < 255) {
		/* multicast */
		i = 1; (void)setsockopt(sfd, IPPROTO_IP, IP_MULTICAST_LOOP, &i, sizeof(i));
		i = 31; (void)setsockopt(sfd, IPPROTO_IP, IP_MULTICAST_TTL, &i, sizeof(i));

		memset(&gamr, 0, sizeof(gamr));
		memcpy(&gamr.imr_multiaddr, &inb, sizeof(in_addr_t));
		memcpy(&gamr.imr_interface, &ina, sizeof(in_addr_t));
		/* join the group */
		(void)setsockopt(sfd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
					 &gamr,sizeof(gamr));
	}
	sockop_broadcast(sfd, 1);
	sockop_reuseaddr(sfd, 1);
	sin.sin_port = htons(0); /* send from any port */
	memcpy(&binda, &sin, sizeof(sin));
	if (bind(sfd, (struct sockaddr *)&sin, sizeof(sin)) == -1) exit(111);

	/* ... and connect to */
	memset(&sendt, 0, sizeof(sendt));
	sendt.sin_family = AF_INET;
	memcpy(&sendt.sin_addr, &ina, sizeof(in_addr_t));
	sendt.sin_port = htons(CLUSTER_PORT);


	for (fd = -1;;) {
		if (fd == -1) {
			fd = socket(AF_INET, SOCK_STREAM, 0);
			if (fd > -1) {
				memcpy(&sin, &binda, sizeof(sin));
				if (bind(fd, (struct sockaddr *)&sin, sizeof(sin)) == -1) {
					close(fd);
					fd = -1;
				}
			}
		}

		do {
			tv.tv_sec = expire_cpipes();
			tv.tv_usec = 0;

			FD_ZERO(&rfds);
			FD_ZERO(&wfds);
			FD_SET(bfd, &rfds);
			maxfd = select_cpipes(bfd, &rfds, &wfds);

			ptv = (tv.tv_sec < 0)
				? ((struct timeval *)0)
				: ((struct timeval *)&tv);
			r = select(maxfd+1, &rfds, &wfds, (fd_set *)0, ptv);
		} while (r == -1 && errno == EINTR);
		if (r < 0) {
			exit(250);
		}
		if (r == 0) continue;

		update_cpipes(&rfds, &wfds);

		if (fd != -1 && !FD_ISSET(bfd, &rfds)) continue;

		i = sizeof(sin);
		r = recvfrom(bfd, packet, sizeof(packet), 0,
				(struct sockaddr *)&sin, &i);
		if (r < CLUSTER_SYNC+4) continue;
		if (memcmp(&sin.sin_addr, &inb, sizeof(inb)) == 0) continue;
		if (memcmp(&sin.sin_addr, &ina, sizeof(ina)) == 0) continue;


		/* copy port number for copy back */
		memcpy(&sin.sin_port, packet+CLUSTER_SYNC, sizeof(sin.sin_port));
		if (!(c = add_cpipe(fd, &sin))) continue;
		if (fd != c->left) {
			/* add_cpipe changed fd; close old one */
			(void)close(fd);
			fd = c->left;
			/* already listening... */
		} else {
			if (listen(fd, 1) == -1) {
				(void)close(fd);
				fd = c->left = c->right = -1;
				continue;
			}
		}

		i = sizeof(sin);
		if (getsockname(fd, (struct sockaddr *)&sin, &i) == -1) {
			(void)close(fd);
			fd = c->left = c->right = -1;
			continue;
		}

		memcpy(packet+CLUSTER_SYNC, &sin.sin_port, sizeof(sin.sin_port));
		if (sendto(sfd, packet, r, 0,
					(struct sockaddr *)&sendt,
					sizeof(sendt)) == -1) {
			(void)close(fd);
			fd = c->left = c->right = -1;
		}
		if (fd > -1) {
			sockop_nonblock(fd, 1);
			sockop_keepalive(fd, 1);
			sockop_ndelay(fd, 1);
			fd = -1;
		}
	}
}

